src/app/convai-checker.component.spec.ts

```diff
@@ -1455,4 +1500,394 @@ describe('Convai checker test', () => {
+    fixture.componentInstance.setDemoSettings(demoSettings);
+
+    fixture.detectChanges();
+    let checker = fixture.componentInstance.checker;
+
+    // Set up the mock responses for the series of three requests that will be
+    // made in the test.
+    let queryTexts = [
+      'Your mother was a hamster',
+      'Your father smelled of elderberries',
+      'What is the air velocity of an unladen swallow?'
+    ];
+
+    let callCount = 0;
+    let mockResponses = [
+      getMockCheckerResponseWithScore(0.9, checker.getToken(queryTexts[0])),
+      getMockCheckerResponseWithScore(0.7, checker.getToken(queryTexts[1])),
+      getMockCheckerResponseWithScore(0.2, checker.getToken(queryTexts[2]))
+    ];
+
+    let mockBackend = TestBed.get(MockBackend);
+    mockBackend.connections
+     .subscribe((connection: MockConnection) => {
+       fixture.detectChanges();
+       // Check the UI state before returning the repsonse.
+
+       verifyEmojiWidgetVisible();
+
+       // TODO(rachelrosen): Figure out how to "fast-forward" the state of the
+       // animation in the test environment without messing up the other test
+       // state, and then uncomment the code below to check the opacity of the
+       // emoji icons here.
+       // verifyEmojiIconsInDomWithZeroOpacity();
+
+       expect(checker.statusWidget.isLoading).toBe(true);
+
+       connection.mockRespond(
+         new Response(
+           new ResponseOptions({
+              body: mockResponses[callCount]
+           })
+         )
+       );
+
+       // Wait for async code to complete.
+       fixture.whenStable().then(() => {
+         fixture.detectChanges();
+         // Checks the UI state after the response has been received.
+
+         // Checks that loading has stopped.
+         expect(checker.statusWidget.isLoading).toBe(false);
+         expect(checker.statusWidget.isPlayingLoadingAnimation).toBe(false);
+
+         verifyEmojiWidgetVisible();
+
+         // For each of the mock responses, the thresholds should indicate a
+         // different emoji.
+         let expectedEmoji: Emoji;
+         if (callCount === 0) {
+           expectedEmoji = Emoji.SAD;
+         } else if (callCount === 1) {
+           expectedEmoji = Emoji.NEUTRAL;
+         } else {
+           expectedEmoji = Emoji.SMILE;
+         }
+         verifyLoadingWidgetHasEmoji(checker, expectedEmoji);
+
+         if (callCount < 2) {
+           callCount++;
+           // Fire another request.
+           setTextAndFireInputEvent(queryTexts[callCount], textArea);
+         }
+       });
+    });
+
+    verifyEmojiWidgetVisible();
+
+    let textArea = fixture.debugElement.query(
+      By.css('#' + checker.inputId)).nativeElement;
+
+    // Send an input event to trigger the service call.
+    setTextAndFireInputEvent(queryTexts[callCount], textArea);
+  }));
+
+  it('Test loading icon style setting change. Circle square diamond to emoji',
```

Comment:
This ends up going through a lot of loops to check each score threshold, I think you could shorten the test a bit and still get the same kind of benefit by just doing one check with shapes and then another check with emojis.  Is there a reason to do three checks of each that I'm missing?

Reply:
Because the animation has a lot of state that can get messed up, I prefer to run these sequential tests because they will identity potential bugs in transitions / animation inconsistencies between checks.